istioctl/pkg/rbac/upgrade.go

```diff
@@ -0,0 +1,296 @@
+
+// WorkloadLabels is the workload labels, for example, app: productpage.
+type WorkloadLabels map[string]string
+
+type Upgrader struct {
+	IstioConfigStore model.ConfigStore
+	K8sClient        *kubernetes.Clientset
+	// RoleToWorkloadLabels maps the ServiceRole name to the map of service name to the workload labels
+	// that the service pointing to.
+	// We could have remove the `service` key layer to make it ServiceRole name maps to workload labels.
+	// However, this is needed for unit tests.
+	RoleToWorkloadLabels map[string]ServiceToWorkloadLabels
+	RbacFiles            []string
+	serviceRoles         []model.Config
+	serviceRoleBindings  []model.Config
+}
+
+const (
+	apiVersionField = "apiVersion: "
+	kindField       = "kind: "
+	metadataField   = "metadata:"
+	nameField       = "name: "
+	namespaceField  = "namespace: "
+	specField       = "spec:"
+
+	newFileNameSuffix = "-v2"
+)
+
+var (
+	configDescriptor = model.ConfigDescriptor{
+		model.ServiceRole,
+		model.ServiceRoleBinding,
+		model.AuthorizationPolicy,
+	}
+)
+
+// UpgradeCRDs is the main function that converts RBAC v1 to v2 for local policy files.
+func (ug *Upgrader) UpgradeCRDs() error {
+	for _, fileName := range ug.RbacFiles {
+		_, err := ug.upgradeLocalFile(fileName)
+		if err != nil {
+			return fmt.Errorf("failed to convert file %s with error %v", fileName, err)
+		}
+	}
+	return nil
+}
+
+// upgradeLocalFile performs upgrade/conversion for all ServiceRole and ServiceRoleBinding in the fileName.
+func (ug *Upgrader) upgradeLocalFile(fileName string) (string, error) {
+	err := ug.createRoleAndBindingLists(fileName)
+	if err != nil {
+		return "", err
+	}
+	newFile, newFilePath, err := checkAndCreateNewFile(fileName)
+	if err != nil {
+		return "", err
+	}
+	for _, serviceRole := range ug.serviceRoles {
+		if err := ug.updateRoleToWorkloadLabels(serviceRole); err != nil {
+			return "", err
+		}
+		role := ug.upgradeServiceRole(serviceRole)
+		err = writeToFile(newFile, role)
+		if err != nil {
+			return "", err
+		}
+	}
+	for _, serviceRoleBinding := range ug.serviceRoleBindings {
+		authzPolicy := ug.createAuthorizationPolicyFromRoleBinding(serviceRoleBinding)
+		err = writeToFile(newFile, authzPolicy)
+		if err != nil {
+			return "", err
+		}
+	}
+	fmt.Println(fmt.Sprintf("New authorization policies have been upgraded to %s", newFilePath))
+	return newFilePath, nil
+}
+
+// writeToFile writes data from `config` to `file`.
+func writeToFile(file *os.File, config model.Config) error {
+	var stringBuilder strings.Builder
+	schema, exists := configDescriptor.GetByType(config.Type)
+	if !exists {
+		return fmt.Errorf("unknown kind %q for %v", crd.ResourceName(config.Type), config.Name)
+	}
+	obj, err := crd.ConvertConfig(schema, config)
+	if err != nil {
+		return fmt.Errorf("could not decode %v: %v", config.Name, err)
+	}
+	stringBuilder.WriteString(fmt.Sprintf("%s%s/%s\n", apiVersionField, config.Group, config.Version))
```

Comment:
I feel this is not the best way to serialize the k8s CRD, there might be a better way that you don't need to output these meta fields manually. And at least if there is not such way (unlikely), you'd better wrap this in a template file instead of using `stringBuilder.WriteString()` manually.

Reply:
If you remember our discussion offline, I told you that `yaml.Marshal` outputted irrelevant metadata and was much different from the output from `convert_ingress.go`. I just found out that `convert_ingress.go` uses `yaml` package at `github.com/ghodss/yaml`, however IntelliJ auto-completes `yaml` using package at `https://gopkg.in/yaml.v2`. Changing the package fixes the problem, so we won't need to use Go template or anything like that.